[SPARK-19342][SPARKR] bug fixed in collect method for collecting timestamp column

titicaca opened the pull request:
## What changes were proposed in this pull request?

Fix a bug in collect method for collecting timestamp column, the bug can be reproduced as shown in the following codes and outputs:

```
library(SparkR)
sparkR.session(master = "local")
df <- data.frame(col1 = c(0, 1, 2), 
                 col2 = c(as.POSIXct("2017-01-01 00:00:01"), NA, as.POSIXct("2017-01-01 12:00:01")))

sdf1 <- createDataFrame(df)
print(dtypes(sdf1))
df1 <- collect(sdf1)
print(lapply(df1, class))

sdf2 <- filter(sdf1, "col1 > 0")
print(dtypes(sdf2))
df2 <- collect(sdf2)
print(lapply(df2, class))
```

As we can see from the printed output, the column type of col2 in df2 is converted to numeric unexpectedly, when NA exists at the top of the column. 

This is caused by method `do.call(c, list)`, if we convert a list, i.e. `do.call(c, list(NA, as.POSIXct("2017-01-01 12:00:01"))`, the class of the result is numeric instead of POSIXct. 

Therefore, we need to cast the data type of the vector explicitly. 



## How was this patch tested?

The patch can be tested manually with the same code above.

titicaca committed:
SPARK-19342 bug fixed in collect method for collecting timestamp column

AmplabJenkins commented:
Can one of the admins verify this patch?

felixcheung commented:
Thanks! I can verify this case and the fix. 
Could you please add some tests for this?

HyukjinKwon commented:
Oh @felixcheung , I was writing a comment but I just saw you. I was looking into this for my curiosity.

Isn't this due to R type coercion rule with POSIXlt? 

```r
> str(data.frame(col1 = c(as.POSIXct("2017-01-01 12:00:01"), NA)))
'data.frame':	2 obs. of  1 variable:
 $ col1: POSIXct, format: "2017-01-01 12:00:01" NA
```

```r
> str(data.frame(col1 = c(NA, as.POSIXct("2017-01-01 12:00:01"))))
'data.frame':	2 obs. of  1 variable:
 $ col1: num  NA 1.48e+09
```

HyukjinKwon commented:
(I might be wrong but was suspecting that it returns `NA` first as `logical` when we collect via `SerDe.scala` and then it ends up `numeric` due to the type coercion when `NA` is located first as above.)

HyukjinKwon commented:
Oh. it was all written in the PR description... I removed my uesless comments.

titicaca committed:
SPARK-19342 bug fixed in collect method for collecting timestamp column

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1136,9 +1136,17 @@ setMethod("collect",
+                      vec <- do.call(c, colTail)
+                      classVal <- class(vec)
+                      vec <- c(rep(NA, valueIndex[1] - 1), vec)
+                      class(vec) <- classVal
```
Hmm, what happened here?
if you want to drop the NA and use the rest to infer the class you can do `col[!is.na(col)]`

titicaca commented:
Sure. Shall I add the tests in pkg/inst/tests/testthat/test_sparkSQL.R?

felixcheung commented:
yes. but please see my other comment

titicaca committed:
SPARK-19342 bug fixed in collect method for collecting timestamp column, if column values are all NAs, the type is logical

titicaca commented:
Sorry for the late reply. I figured out that the tests failed because if a vector is with only NAs, the type is logical, therefore we cannot cast the type in that case. I have updated the codes and added some tests for that. Thank you for the advice.

felixcheung commented:
great! @shivaram could you get Jenkins to test this fix please? I don't seem to have the power to command it :)

shivaram commented:
Jenkins, retest this please

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1138,6 +1138,9 @@ setMethod("collect",
                   if (!is.null(PRIMITIVE_TYPES[[colType]]) && colType != "binary") {
                     vec <- do.call(c, col)
                     stopifnot(class(vec) != "list")
+                    # If vec is an vector with only NAs, the type is logical
```
if the DataFrame column is of type string, shouldn't it converts to R as character (which can be all NA), even though the column only has NULL (which maps to NA in R)?

it seems with this change it would become logical in R instead of character.

titicaca replied:
Yes. My first commit was trying to cast the column to its corresponding R data type explicitly, even if it is an vector with all NAs. However some existed tests were failed and expecting to get logical NA. For example
```
3. Failure: column functions (@test_sparkSQL.R#1280) ---------------------------
collect(select(df, first(df$age)))[[1]] not equal to NA.
Types not compatible: double vs logical
4. Failure: column functions (@test_sparkSQL.R#1282) ---------------------------
collect(select(df, first("age")))[[1]] not equal to NA.
Types not compatible: double vs logical
```

titicaca replied:
In local R, if we try
```
df <- data.frame(x = c(0,1,2), y = c(NA, NA, 1))
class(head(df, 1)$y)
```
The output is still numeric instead of logical. But the existed test is expecting NA logical instead of NA numeric.

So is it necessary to correct the existed tests, for example @test_sparkSQL.R#1280
from `expect_equal(collect(select(df, first(df$age)))[[1]], NA)` to
`expect_equal(collect(select(df, first(df$age)))[[1]], NA_real_)`

shivaram commented:
Jenkins, ok to test

SparkQA commented:
**[Test build #71969 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71969/testReport)** for PR 16689 at commit [`6a0eb3f`](https://github.com/apache/spark/commit/6a0eb3f8789b6a66a4d1419c389fdda2edc0bc95).

SparkQA commented:
**[Test build #71969 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71969/testReport)** for PR 16689 at commit [`6a0eb3f`](https://github.com/apache/spark/commit/6a0eb3f8789b6a66a4d1419c389fdda2edc0bc95).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71969/
Test FAILed.

titicaca committed:
SPARK-19342 fix R code style error in test

titicaca committed:
SPARK-19342 If vec is an vector with only NAs, the type will be cast to PRIMITIVE_TYPE, in addition two existed tests (@test_sparkSQL.R#1280 and @test_sparkSQL.R#1282) are modfied

SparkQA commented:
**[Test build #71982 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71982/testReport)** for PR 16689 at commit [`43e334a`](https://github.com/apache/spark/commit/43e334a1debc31ab4383589ffd4144cc9d1e7633).

SparkQA commented:
**[Test build #71982 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71982/testReport)** for PR 16689 at commit [`43e334a`](https://github.com/apache/spark/commit/43e334a1debc31ab4383589ffd4144cc9d1e7633).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71982/
Test FAILed.

titicaca commented:
I have modified the codes and tests, including the existed tests @test_sparkSQL.R#1280 and @test_sparkSQL.R#1282. 

Like in local R,  now NA column of the SparkDataFrame will also be collected as its corresponding type instead of logical NA.

felixcheung commented:
Just to make sure you see this: https://github.com/apache/spark/pull/16689#issuecomment-275063425

titicaca committed:
SPARK-19342 fix for R style checks

SparkQA commented:
**[Test build #72182 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72182/testReport)** for PR 16689 at commit [`7903bb3`](https://github.com/apache/spark/commit/7903bb3cc44477d9f7a25971ef4487af0627d333).

SparkQA commented:
**[Test build #72182 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72182/testReport)** for PR 16689 at commit [`7903bb3`](https://github.com/apache/spark/commit/7903bb3cc44477d9f7a25971ef4487af0627d333).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72182/
Test FAILed.

titicaca committed:
SPARK-19342 remove the comment

SparkQA commented:
**[Test build #72186 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72186/testReport)** for PR 16689 at commit [`8379c38`](https://github.com/apache/spark/commit/8379c3834fc27e3303501536181bd85372493982).

SparkQA commented:
**[Test build #72186 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72186/testReport)** for PR 16689 at commit [`8379c38`](https://github.com/apache/spark/commit/8379c3834fc27e3303501536181bd85372493982).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72186/
Test PASSed.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1138,6 +1138,11 @@ setMethod("collect",
                     stopifnot(class(vec) != "list")
+                    class(vec) <-
+                      if (colType == "timestamp")
+                        c("POSIXct", "POSIXt")
```
why should the class be `c("POSIXct", "POSIXt")` in this case?

titicaca replied:
Because `PRIMITIVE_TYPES[["timestamp"]]` is POSIXct, it usually comes with POSIXt together. POSIXt is virtual class used to allow operations such as subtraction to mix the two classes POSIXct and POSIXlt.
The previous convertion will also convert timestamp to c("POSIXct", "POSIXt").

felixcheung replied:
Should `PRIMITIVE_TYPES[["timestamp"]]` be changed then
https://github.com/apache/spark/blob/master/R/pkg/R/types.R#L32

titicaca replied:
It looks better if it won't affect other methods. I will try it. Thanks for the advice.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1138,6 +1138,11 @@ setMethod("collect",
+                      if (colType == "timestamp")
+                        c("POSIXct", "POSIXt")
+                      else
+                        PRIMITIVE_TYPES[[colType]]
```
by setting these instead of having it inferred - does this break any existing behavior? does any type differ because of this line of change?

titicaca replied:
Currently all tests are passed, except for the two modified tests with NA types as discussed before.  The followings are the all type convertions from SparkDataframe to R data.frame, which have been tested in the existing tests in test_sparkSQL.R. 
```
PRIMITIVE_TYPES <- as.environment(list(
  "tinyint" = "integer",
  "smallint" = "integer",
  "int" = "integer",
  "bigint" = "numeric",
  "float" = "numeric",
  "double" = "numeric",
  "decimal" = "numeric",
  "string" = "character",
  "binary" = "raw",
  "boolean" = "logical",
  "timestamp" = "POSIXct",
  "date" = "Date",
  # following types are not SQL types returned by dtypes(). They are listed here for usage
  # by checkType() in schema.R.
  # TODO: refactor checkType() in schema.R.
  "byte" = "integer",
  "integer" = "integer"
  ))
```

titicaca committed:
SPARK-19342 modify PRIMITIVE_TYPES for type timestamp

SparkQA commented:
**[Test build #72250 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72250/testReport)** for PR 16689 at commit [`407c625`](https://github.com/apache/spark/commit/407c6254246be82e88098aa08838284e6861838a).

SparkQA commented:
**[Test build #72250 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72250/testReport)** for PR 16689 at commit [`407c625`](https://github.com/apache/spark/commit/407c6254246be82e88098aa08838284e6861838a).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72250/
Test FAILed.

titicaca commented:
I tried to modify the PRIMITIVE_TYPES for timestamp, but it had a side effect on coltypes method.

In test_sparkSQL.R#2262, `expect_equal(coltypes(DF), c("integer", "logical", "POSIXct"))`, coltypes return a list instead of a vector because of the convertion from timestamp to `c(POSIXct, POSIXt)`

felixcheung commented:
hmm, that's not a super big issue since vector and list is more or less the same in R.
I think it might be better if we are treating the type consistently, although it might be a concerning if this is changing in a non-backward compatible manner.

let me try to find some time to test this out? thanks!

titicaca committed:
SPARK-19342 fix for method coltypes in test_sparkSQL.R#2262

SparkQA commented:
**[Test build #72378 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72378/testReport)** for PR 16689 at commit [`d6d454e`](https://github.com/apache/spark/commit/d6d454ec0a587c456d5e4a964784589643aa4730).

SparkQA commented:
**[Test build #72378 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72378/testReport)** for PR 16689 at commit [`d6d454e`](https://github.com/apache/spark/commit/d6d454ec0a587c456d5e4a964784589643aa4730).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72378/
Test PASSed.

titicaca commented:
Thanks. I tried to fix the method `coltypes` for the modification of the timestamp, and it can pass all the tests now.

felixcheung commented:
hmm, this seems like a reasonable approach. With these changes:
- collect on timestamp would get `c("POSIXct", "POSIXt")`
- coltypes output will not change

@shivaram what do you think?

shivaram commented:
@felixcheung @titicaca Just to make sure I understand, collect on timestamp was getting `c("POSIXct", "POSIXt")` even before this change ?

titicaca commented:
Yes, collect on timestamp was getting `c("POSIXct", "POSIXt")`. But when NA exists at the top of the timetamp column, it was getting `numeric` as I described in the PR description.

shivaram commented:
Ok - I think this sounds good then ! @felixcheung Let me know if you want me to take a look at the code as well or if not feel free to merge when you think its ready

felixcheung commented:
great, this is a good catch and thank you for fixing this @titicaca 
merging to master, branch-2.1

felixcheung commented:
@titicaca do you have a JIRA id on https://issues.apache.org? We would resolve the bug to you.

titicaca commented:
Yes. The JIRA id is SPARK-19342. Thank you for the help and advices :)

srowen commented:
@titicaca he means, what is your user ID on JIRA? so we can credit you. It's clear what the JIRA is.

titicaca commented:
Thanks for the reminder. I may have forgotten to mention that I am the reporter of this JIRA bug. My JIRA ID is also titicaca. Thank you!

felixcheung commented:
Great, done!
Looking forward to more contributions from you :)
